tools/slang-cpp-extractor/slang-cpp-extractor-main.cpp

```diff
@@ -0,0 +1,1944 @@
+
+protected:
+    static Node::Type _toNodeType(IdentifierStyle style);
+
+    bool _isMarker(const UnownedStringSlice& name)
+    {
+        return name.startsWith(m_options->m_prefixMark.getUnownedSlice()) && name.endsWith(m_options->m_postfixMark.getUnownedSlice());
+    }
+
+    SlangResult _maybeParseNode(Node::Type type);
+    SlangResult _maybeParseField();
+
+    SlangResult _maybeParseType(UnownedStringSlice& outType);
+
+    SlangResult _maybeParseType(UnownedStringSlice& outType, Index& ioTemplateDepth);
+    SlangResult _maybeParseTemplateArgs(Index& ioTemplateDepth);
+    SlangResult _maybeParseTemplateArg(Index& ioTemplateDepth);
+
+    SlangResult _calcDerivedTypesRec(Node* node);
+
+    void _consumeTypeModifiers();
+
+    SlangResult _consumeToSync();
+
+    TokenList m_tokenList;
+    TokenReader m_reader;
+
+    Node* m_currentNode;            ///< The current scope being processed
+
+    RefPtr<Node> m_rootNode;        ///< The root scope 
+
+    List<Node*> m_baseTypes;        ///< All of the types which are base. Only set after calcDerivedTypes
+
+    DiagnosticSink* m_sink;
+
+    NamePool* m_namePool;
+
+    const Options* m_options;
+
+    IdentifierLookup* m_identifierLookup;
+    StringSlicePool* m_typePool;
+};
+
+CPPExtractor::CPPExtractor(StringSlicePool* typePool, NamePool* namePool, DiagnosticSink* sink, IdentifierLookup* identifierLookup):
+    m_typePool(typePool),
+    m_sink(sink),
+    m_namePool(namePool),
+    m_identifierLookup(identifierLookup)
+{
+    m_rootNode = new Node(Node::Type::Namespace);
+}
+
+SlangResult CPPExtractor::expect(TokenType type, Token* outToken)
+{
+    if (m_reader.peekTokenType() != type)
+    {
+        m_sink->diagnose(m_reader.peekToken(), CPPDiagnostics::expectingToken, type);
+        return SLANG_FAIL;
+    }
+
+    if (outToken)
+    {
+        *outToken = m_reader.advanceToken();
+    }
+    else
+    {
+        m_reader.advanceToken();
+    }
+    return SLANG_OK;
+}
+
+bool CPPExtractor::advanceIfToken(TokenType type, Token* outToken)
+{
+    if (m_reader.peekTokenType() == type)
+    {
+        Token token = m_reader.advanceToken();
+        if (outToken)
+        {
+            *outToken = token;
+        }
+        return true;
+    }
+    return false;
+}
+
+bool CPPExtractor::advanceIfMarker(Token* outToken)
+{
+    const Token peekToken = m_reader.peekToken();
+    if (peekToken.type == TokenType::Identifier && _isMarker(peekToken.Content))
+    {
+        m_reader.advanceToken();
+        if (outToken)
+        {
+            *outToken = peekToken;
+        }
+        return true;
+    }
+    return false;
+}
+
+bool CPPExtractor::advanceIfStyle(IdentifierStyle style, Token* outToken)
+{
+    if (m_reader.peekTokenType() == TokenType::Identifier)
+    {
+        IdentifierStyle readStyle = m_identifierLookup->get(m_reader.peekToken().Content);
+        if (readStyle == style)
```

Comment:
One note is that we *might* be able to avoid having a context-sensitive parse in the C++ extractor by having it rely on a combination of heuristic rules and convention to disambiguate the interesting cases. The worry with using a "proper" context-sensitive parse is that you can never really have all of the context without processing `#include` directives to discover other user-defined names.